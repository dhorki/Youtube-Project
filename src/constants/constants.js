export const styles = {
  colors: {
    light: {
      appIcon: '#ff0200',
      background: '#F9F9F9',
      barBackground: '#ffffff',
      primary: '#030303',
      secondary: '#606060',
      linkActive: '#ff0200',
      linkHoverBackground: '#ececec',
      inputBackground: '#ffffff',
      inputBorder: '#cccccc',
      searchButtonBackground: '#f8f8f8',
      // google button
      white: '#fff',
      googleBlue: '#4285f4',
      buttonActiveBlue: '#1669f2',
    },
    dark: {
      appIcon: '#ff0200',
      background: '#181818',
      barBackground: '#212121',
      primary: '#ffffff',
      secondary: '#aaaaaa',
      linkActive: '#ff0200',
      linkHoverBackground: '#303030',
      inputBackground: '#121212',
      inputBorder: '#303030',
      searchButtonBackground: '#323232',
      // google button
      white: '#fff',
      googleBlue: '#4285f4',
      buttonActiveBlue: '#1669f2',
    },
  },
  props: {
    transitionTime: '0.4s',
    sidebarWidth: '150px',
  },
};

export const screenSizes = {
  xsmall: '675px',
  small: '830px',
  mid: '1200px',
};

export const mainViewTypes = {
  searchGallery: 'searchGallery',
  favoritesGallery: 'favoritesGallery',
  videoDetails: 'videoDetails',
};

export const environmentActions = {
  setTheme: 'setTheme',
  setSidebarShow: 'setSidebarShow',
  setSearchQuery: 'setSearchQuery',
  setModalShow: 'setModalShow',
  setUser: 'setUser',
};

export const modalTypes = {
  login: 'login',
  register: 'register',
  resetPassword: 'reset-password',
  none: 'none',
};

export const resetSearchQueryAction = {
  type: environmentActions.setSearchQuery,
  payload: {
    q: '',
  },
};

export const hideModalAction = {
  type: environmentActions.setModalShow,
  payload: {
    modalShow: modalTypes.none,
  },
};
