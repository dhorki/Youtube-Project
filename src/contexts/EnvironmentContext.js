import { createContext } from 'react';

export const EnvironmentContext = createContext();
