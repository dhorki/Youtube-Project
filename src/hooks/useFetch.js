import { useEffect, useRef, useState } from 'react';

const cache = { current: {} };

export const useFetch = (url, cacheIt = true) => {
  const _isMounted = useRef(true); // Initial value _isMounted = true
  const [data, setData] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);

  useEffect(() => {
    let controller = new AbortController();
    // console.log(cache);

    const fetchData = async () => {
      setLoading(true);
      const cachedResult = cache.current[url];
      if (cachedResult && !cachedResult.error) {
        const result = cache.current[url];
        // console.log('fetching from cache ' + url);
        setData(result);
      } else {
        try {
          // console.log('fetching');
          const response = await fetch(url, { signal: controller.signal });
          const result = await response.json();
          if (result.error) {
            setError(result.error);
          }
          if (cacheIt) {
            cache.current[url] = { ...result, fromCache: true }; // set response in cache;
          }
          setData(result);
        } catch (e) {
          // console.log(e);
          if (_isMounted.current) {
            setError(e);
          }
        }
      } //else
      if (_isMounted.current) {
        setLoading(false);
      }
    };
    fetchData();

    return () => {
      _isMounted.current = false;
      controller.abort();
    };
  }, [url, cacheIt]);

  return {
    data,
    loading,
    error,
  };
};
