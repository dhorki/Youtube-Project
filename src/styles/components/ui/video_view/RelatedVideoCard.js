import { Link } from 'react-router-dom';
import styled from 'styled-components';
import { styles } from '../../../../constants/constants';

export const StyledRelatedVideoCard = styled(Link)`
  display: flex;
  flex-direction: row;
  width: 360px;

  margin-left: 5px;
  margin-bottom: 10px;
  background-color: ${({ theme }) => theme.background};
  color: ${({ theme }) => theme.primary};
  text-decoration: none;

  transition: background-color ${styles.props.transitionTime} ease;

  > img {
    width: 168px;
    height: 94px;
  }

  > div {
    display: flex;
    flex-direction: column;
    width: 226px;
    margin-left: 10px;

    > .videocard-title {
      color: ${({ theme }) => theme.primary};
      font-size: 14px;
      font-weight: bold;
    }

    > .videocard-channel-title {
      margin-top: 10px;
      color: ${({ theme }) => theme.secondary};
      font-size: 14px;
    }
  }
`;
