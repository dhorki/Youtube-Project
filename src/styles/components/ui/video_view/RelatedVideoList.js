import styled from 'styled-components';
import { screenSizes } from '../../../../constants/constants';

export const StyledRelatedVideoList = styled.div`
  display: flex;
  flex-direction: column;
  flex-wrap: wrap;
  justify-content: center;
  align-items: center;
  height: 100%;
  width: 390px;

  @media only screen and (max-width: ${screenSizes.mid}) {
    width: 100%;
    flex-direction: row;
    align-items: top;
  }

  > h3 {
    color: ${({ theme }) => theme.primary};
    width: 100%;
    text-align: center;
  }
`;
