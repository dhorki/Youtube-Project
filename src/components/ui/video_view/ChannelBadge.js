import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import { StyledChannelBadge } from '../../../styles/components/ui/video_view/ChannelBadge';
import { useFetch } from '../../../hooks/useFetch';
import { intToMagnitude } from '../../../helpers/formatters';
import { EnvironmentContext } from '../../../contexts/EnvironmentContext';

export const ChannelBadge = ({ id }) => {
  const { environment } = useContext(EnvironmentContext);
  const { theme } = environment;

  const url =
    'https://www.googleapis.com/youtube/v3/channels?' +
    `key=${process.env.REACT_APP_YOUTUBE_API_KEY}&` +
    'part=snippet,statistics&' +
    `id=${id}&` +
    'maxResults=24';

  const { data, loading, error } = useFetch(url);

  let image = '',
    title = '#?',
    subscriberCount = 0;
  if (!loading && !error) {
    const item = data?.items[0];
    image = item.snippet.thumbnails.default.url;
    title = item.snippet.title;
    subscriberCount = item.statistics.subscriberCount;
  }

  return (
    <StyledChannelBadge theme={theme}>
      {loading ? null : (
        <div className="badge">
          {!error ? <img src={image} alt={title} /> : null}
          <div>
            <p className="title">{title}</p>
            <p className="subscribers">{intToMagnitude(subscriberCount)} subscribers</p>
          </div>
        </div>
      )}
    </StyledChannelBadge>
  );
};

ChannelBadge.propTypes = {
  id: PropTypes.string.isRequired,
};
