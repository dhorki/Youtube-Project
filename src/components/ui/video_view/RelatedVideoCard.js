import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import { timeSince } from '../../../helpers/formatters';
import { decodeHtml } from '../../../helpers/decodeHtml';
import { StyledRelatedVideoCard } from '../../../styles/components/ui/video_view/RelatedVideoCard';
import { EnvironmentContext } from '../../../contexts/EnvironmentContext';

export const RelatedVideoCard = ({ video }) => {
  const { environment } = useContext(EnvironmentContext);
  const { theme } = environment;

  const { id, snippet } = video;
  const { videoId } = id;
  const { title, channelTitle, thumbnails, publishTime } = snippet;
  const { url } = thumbnails.medium;
  const d = new Date(publishTime);

  const upperCaseCharsCount = title.length - title.replaceAll(/[A-Z]/g, '').length;
  const limitLength = upperCaseCharsCount > 50 ? 50 : upperCaseCharsCount > 30 ? 60 : 70;

  const fixedTitle =
    title.length >= limitLength ? title.slice(0, limitLength) + '...' : title;

  return (
    <StyledRelatedVideoCard
      className="animate__animated animate__fadeIn"
      theme={theme}
      to={`/${videoId}`}
    >
      <img src={url} alt="video" />
      <div>
        <div className="videocard-title">{decodeHtml(fixedTitle)}</div>
        <div className="videocard-channel-title">
          {decodeHtml(channelTitle)}
          <br />
          {timeSince(d)} ago.
        </div>
      </div>
    </StyledRelatedVideoCard>
  );
};

RelatedVideoCard.propTypes = {
  video: PropTypes.object.isRequired,
};
