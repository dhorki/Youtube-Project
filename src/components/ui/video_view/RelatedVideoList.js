import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import { StyledRelatedVideoList } from '../../../styles/components/ui/video_view/RelatedVideoList';
import { useFetch } from '../../../hooks/useFetch';
import { RelatedVideoCard } from './RelatedVideoCard';
import { LoadingAnimation } from '../loading/LoadingAnimation';
import { ErrorText } from '../error/ErrorText';
import { EnvironmentContext } from '../../../contexts/EnvironmentContext';

export const RelatedVideoList = ({ relatedTo }) => {
  const { environment } = useContext(EnvironmentContext);
  const { theme } = environment;

  const url =
    'https://www.googleapis.com/youtube/v3/search?' +
    `key=${process.env.REACT_APP_YOUTUBE_API_KEY}&` +
    'part=snippet&' +
    'type=video&' +
    'order=relevance&' +
    `relatedToVideoId=${relatedTo}&` +
    'maxResults=24';

  const { data, loading, error } = useFetch(url);
  // const { data, loading, error } = { data: [], loading: true, error: null };

  let filteredItems = [];

  if (!loading && !error) {
    filteredItems = data.items?.filter(
      (item) => item.id.kind.includes('youtube#video') && item.snippet
    );
  }

  return (
    <StyledRelatedVideoList theme={theme}>
      <h3>Related Videos</h3>
      {loading ? (
        <LoadingAnimation />
      ) : error ? (
        <ErrorText error={error} />
      ) : !filteredItems || filteredItems.length === 0 ? (
        'No Related videos.'
      ) : (
        filteredItems.map((video) => (
          <RelatedVideoCard key={video.id.videoId} video={video} />
        ))
      )}
    </StyledRelatedVideoList>
  );
};

RelatedVideoList.propTypes = {
  relatedTo: PropTypes.string.isRequired,
};
