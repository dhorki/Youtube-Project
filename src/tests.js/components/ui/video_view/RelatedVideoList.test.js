import React from 'react';
import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import { styles } from '../../../../constants/constants';
import { EnvironmentContext } from '../../../../contexts/EnvironmentContext';
import { useFetch } from '../../../../hooks/useFetch';
import { RelatedVideoList } from '../../../../components/ui/video_view/RelatedVideoList';
import { RelatedVideoCard } from '../../../../components/ui/video_view/RelatedVideoCard';
jest.mock('../../../../hooks/useFetch');
jest.mock('../../../../components/ui/video_view/RelatedVideoCard');

describe('RelatedVideoList', () => {
  const environment = {
    theme: styles.colors.light,
    sidebarShow: false,
    q: '',
  };

  it('should render correctly while loading', () => {
    const videoId = 'channelId';

    useFetch.mockReturnValue({
      data: {},
      loading: true,
      error: null,
    });

    const wrapper = mount(
      <EnvironmentContext.Provider value={{ environment }}>
        <RelatedVideoList relatedTo={videoId} />
      </EnvironmentContext.Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly with success fetch', () => {
    const videoId = 'channelId';

    RelatedVideoCard.mockImplementation(({ video }) => (
      <div>RelatedVideoCard {video.id.videoId}</div>
    ));

    useFetch.mockReturnValue({
      data: {
        items: [
          {
            id: {
              videoId: 'videoId1',
              kind: 'youtube#video',
            },
            snippet: {},
          },
          {
            id: {
              videoId: 'videoId2',
              kind: 'youtube#video',
            },
            snippet: {},
          },
          {
            id: {
              videoId: 'videoId3',
              kind: 'youtube#video',
            },
            snippet: {},
          },
        ],
      },
      loading: false,
      error: null,
    });

    const wrapper = mount(
      <EnvironmentContext.Provider value={{ environment }}>
        <RelatedVideoList relatedTo={videoId} />
      </EnvironmentContext.Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly with success fetch', () => {
    const videoId = 'channelId';

    RelatedVideoCard.mockImplementation(({ video }) => (
      <div>RelatedVideoCard {video.id.videoId}</div>
    ));

    useFetch.mockReturnValue({
      data: {
        items: [
          {
            id: {
              videoId: 'videoId1',
              kind: 'youtube#video',
            },
            snippet: {},
          },
          {
            id: {
              videoId: 'videoId2',
              kind: 'youtube#video',
            },
            snippet: {},
          },
          {
            id: {
              videoId: 'videoId3',
              kind: 'youtube#video',
            },
            snippet: {},
          },
        ],
      },
      loading: false,
      error: null,
    });

    const wrapper = mount(
      <EnvironmentContext.Provider value={{ environment }}>
        <RelatedVideoList relatedTo={videoId} />
      </EnvironmentContext.Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly with no related videos', () => {
    const videoId = 'channelId';

    RelatedVideoCard.mockImplementation(({ video }) => (
      <div>RelatedVideoCard {video.id.videoId}</div>
    ));
    useFetch.mockReturnValue({
      data: {
        items: [],
      },
      loading: false,
      error: null,
    });

    const wrapper = mount(
      <EnvironmentContext.Provider value={{ environment }}>
        <RelatedVideoList relatedTo={videoId} />
      </EnvironmentContext.Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly with fail fetch', () => {
    const videoId = 'videoId';

    useFetch.mockReturnValue({
      data: {},
      loading: false,
      error: { error: 'error' },
    });

    const wrapper = mount(
      <EnvironmentContext.Provider value={{ environment }}>
        <RelatedVideoList relatedTo={videoId} />
      </EnvironmentContext.Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
