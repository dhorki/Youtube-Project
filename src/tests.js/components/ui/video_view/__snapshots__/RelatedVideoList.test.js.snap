// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelatedVideoList should render correctly while loading 1`] = `
<div
  className="sc-bczRLJ cSUjCY"
>
  <h3>
    Related Videos
  </h3>
  <div
    className="sc-gsnTZi gEmEgX"
  >
    <img
      alt="loading"
      src="https://i.pinimg.com/originals/d4/2b/b2/d42bb2f00e1042c990b25f90ada9b0ab.gif"
    />
  </div>
</div>
`;

exports[`RelatedVideoList should render correctly with fail fetch 1`] = `
<div
  className="sc-bczRLJ cSUjCY"
>
  <h3>
    Related Videos
  </h3>
  <p
    className="animate__animated animate__fadeIn"
  >
    Error: 
    undefined
  </p>
</div>
`;

exports[`RelatedVideoList should render correctly with no related videos 1`] = `
<div
  className="sc-bczRLJ cSUjCY"
>
  <h3>
    Related Videos
  </h3>
  No Related videos.
</div>
`;

exports[`RelatedVideoList should render correctly with success fetch 1`] = `
<div
  className="sc-bczRLJ cSUjCY"
>
  <h3>
    Related Videos
  </h3>
  <div>
    RelatedVideoCard 
    videoId1
  </div>
  <div>
    RelatedVideoCard 
    videoId2
  </div>
  <div>
    RelatedVideoCard 
    videoId3
  </div>
</div>
`;

exports[`RelatedVideoList should render correctly with success fetch 2`] = `
<div
  className="sc-bczRLJ cSUjCY"
>
  <h3>
    Related Videos
  </h3>
  <div>
    RelatedVideoCard 
    videoId1
  </div>
  <div>
    RelatedVideoCard 
    videoId2
  </div>
  <div>
    RelatedVideoCard 
    videoId3
  </div>
</div>
`;
